File: 6th Semester/Programming Paradigms/NP2 Exercises - Part 2/ex20.py
# No programa da questão 19, o que acontece se digitarmos 0.001 no programa anterior? Caso ele não
# funcione, altere-o de forma a corrigir esse problema.

def fatorador(valor: float, cedulas: list):
    valor += 0.0005 # Retira aproximação ruim
    moedas: bool = True if valor < 1 else False
    aux = int(valor * 100) if moedas else int(valor)
    for cedula in cedulas:
        if (aux // cedula > 0):
            if moedas:
                print(f"{aux//cedula:.0f} moedas de R${(cedula/100):.2f}")
            else:
                print(f"{aux//cedula:.0f} cédulas de R${cedula},00")
            aux = aux % cedula
        if aux <= 0:
            break

valor = float(input('Digite um valor em R$'))
print('O valor informado pode ser pago em ')
if valor > 0 and valor < 0.01:
    print('1 moeda de R$0.01')
elif valor <= 0:
    print('0 cédulas')
else:
    cedulas = [50, 20, 10, 4, 1]
    moedas = [50, 10, 5, 2, 1]
    fatorador(valor, cedulas)
    fatorador(valor-int(valor), moedas)

File: 5th Semester/Programming Paradigms/NP2 Exercises - Part 2/ex20.c
// No programa da questão 19, o que acontece se digitarmos 0.001 no programa anterior? Caso ele não
// funcione, altere-o de forma a corrigir esse problema.

// – Modifique o programa da questão 16 para aceitar valores decimais, ou seja, também contar moedas de
// R$0.01, R$0.02, R$0.05, R$0.10, e R$0.50.

#include <stdio.h>

void fatorador(float valor, int cedulas[], int sizeArray)
{
    valor += 0.0005; // Retira aproximação ruim
    int moedas = valor < 1 ? 1 : 0;
    int aux = moedas ? (int)(valor * 100) : (int)valor;
    for (int i = 0; i < sizeArray && aux > 0; i++)
    {
        if (aux / cedulas[i] > 0)
        {
            if (moedas)
                printf("\n%d moedas de R$%.2f", aux / cedulas[i], (float)cedulas[i] / 100);
            else
                printf("\n%d cédulas de R$%d,00", aux / cedulas[i], cedulas[i]);
            aux = aux % cedulas[i];
        }
    }
}

int main()
{
    float valor;

    printf("Digite um valor em R$");
    scanf("%f", &valor);

    printf("O valor informado pode ser pago em ");
    if (valor > 0 && valor < 0.01)
        printf("1 moeda de R$0.01");
    else if (valor <= 0)
        printf("0 cédulas");
    else
    {
        int cedulas[] = {50, 20, 10, 4, 1};
        int moedas[] = {50, 10, 5, 2, 1};

        fatorador(valor, cedulas, 5);
        fatorador(valor - (int)valor, moedas, 5);
    }

    return 0;
}